daft/iceberg/iceberg_write.py

```diff
@@ -0,0 +1,180 @@
+import warnings
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple
+
+from daft import Series
+from daft.table import MicroPartition
+
+if TYPE_CHECKING:
+    import pyarrow as pa
+    from pyiceberg.io.pyarrow import _TablePartition
+    from pyiceberg.partitioning import PartitionSpec as IcebergPartitionSpec
+    from pyiceberg.schema import Schema as IcebergSchema
+    from pyiceberg.typedef import Record
+
+
+def _coerce_pyarrow_table_to_schema(pa_table: "pa.Table", schema: "IcebergSchema") -> "pa.Table":
+    """Coerces a PyArrow table to the supplied schema
+
+    1. For each field in `pa_table`, cast it to the field in `input_schema` if one with a matching name
+        is available
+    2. Reorder the fields in the casted table to the supplied schema, dropping any fields in `pa_table`
+        that do not exist in the supplied schema
+    3. If any fields in the supplied schema are not present, add a null array of the correct type
+
+    This ensures that we populate field_id for iceberg as well as fill in null values where needed
+    This might break for nested fields with large_strings
+    we should test that behavior
+
+    Args:
+        pa_table (pa.Table): Table to coerce
+        schema (IcebergSchema): PyIceberg schema to coerce to
+
+    Returns:
+        pa.Table: Table with schema == `input_schema`
+    """
+    import pyarrow as pa
+    from pyiceberg.io.pyarrow import schema_to_pyarrow
+
+    input_schema = schema_to_pyarrow(schema)
+
+    input_schema_names = set(input_schema.names)
+
+    # Perform casting of types to provided schema's types
+    cast_to_schema = [
+        (input_schema.field(inferred_field.name) if inferred_field.name in input_schema_names else inferred_field)
+        for inferred_field in pa_table.schema
+    ]
+    casted_table = pa_table.cast(pa.schema(cast_to_schema))
+
+    # Reorder and pad columns with a null column where necessary
+    pa_table_column_names = set(casted_table.column_names)
+    columns = []
+    for name in input_schema.names:
+        if name in pa_table_column_names:
+            columns.append(casted_table[name])
+        else:
+            columns.append(pa.nulls(len(casted_table), type=input_schema.field(name).type))
+    return pa.table(columns, schema=input_schema)
+
+
+def _determine_partitions(
+    spec: "IcebergPartitionSpec", schema: "IcebergSchema", arrow_table: "pa.Table"
+) -> List["_TablePartition"]:
+    """Based on https://github.com/apache/iceberg-python/blob/d8d509ff1bc33040b9f6c90c28ee47ac7437945d/pyiceberg/io/pyarrow.py#L2669"""
```

Comment:
Let's add a note also on why we had to duplicate the code? Namely to support non-temporal transforms until PyIceberg itself can do so.

Also let's find the corresponding issue on the PyIceberg side so that we can be notified when they finally support this.

Reply:
Noting also that it might be possible to change this logic up on our end to make this iterator/streaming-based, which will be more memory efficient?

That way we might also be able to avoid relying on `_get_table_partitions` which is a private method and may break in the future.